File: document/doc/reference/operation/open_simplex_noise/example-1/open_simplex_noise.cpp
#include "lue/framework/algorithm/value_policies/open_simplex_noise.hpp"
#include "lue/framework/algorithm/create_partitioned_array.hpp"
#include "lue/framework/algorithm/value_policies/cast.hpp"
#include "lue/framework/algorithm/value_policies/cell_index.hpp"
#include "lue/document.hpp"
#include "lue/framework.hpp"
#include <hpx/hpx_main.hpp>


class Example: public lue::document::Example
{

    public:

        Example(int const argc, char const* const* argv):

            lue::document::Example{
                [&]() -> cxxopts::Options
                {
                    cxxopts::Options options{std::filesystem::path{argv[0]}.filename()};

                    options.positional_help("<result>");
                    options.add_options()
                        // clang-format off
                        ("result_array", "array to write", cxxopts::value<std::string>())
                        // clang-format on
                        ;
                    options.parse_positional({"result_array"});

                    return options;
                }(),
                argc,
                argv}

        {
        }

    private:

        auto run_implementation() -> int override
        {
            // Command line arguments
            if (!argument_parsed("result_array"))
            {
                throw std::runtime_error("Wrong usage, pass '--help' for more information");
            }

            using namespace lue;
            using namespace lue::value_policies;

            auto const result_array_pathname = argument<std::string>("result_array");

            // [example
            Rank const rank{2};
            using Shape = Shape<Count, rank>;
            using FloatElement = FloatingPointElement<0>;
            using BooleanArray = PartitionedArray<BooleanElement, rank>;
            using FloatArray = PartitionedArray<FloatElement, rank>;

            Shape const array_shape{600, 400};
            BooleanArray const condition_array = create_partitioned_array<BooleanElement>(array_shape, 1);
            FloatArray const x_coordinates = cast<FloatElement>(cell_index<IndexElement>(condition_array, 1));
            FloatArray const y_coordinates = cast<FloatElement>(cell_index<IndexElement>(condition_array, 0));
            int const seed = 5;
            FloatArray const result = open_simplex_noise(x_coordinates, y_coordinates, seed);

            to_gdal(result, result_array_pathname);
            // example]

            return EXIT_SUCCESS;
        }
};


auto main(int argc, char** argv) -> int
{
    return Example{argc, argv}.run();
}

File: document/documentation/reference/operation/local/open_simplex_noise/example-1/open_simplex_noise.py
import sys
from pathlib import Path

import docopt

import lue.framework_x as lfrx
from lue import __version__ as version


@lfrx.runtime_scope
def main() -> None:
    command = Path(sys.argv[0]).name
    usage = f"""\
Usage:
    {command} <result>

Options:
    -h --help   Show this screen and exit
    --version   Show version and exit
    <result>    Array to write
"""
    arguments = docopt.docopt(usage, sys.argv[1:], version=version)
    result_array_path = Path(arguments["<result>"])

    # [example
    boolean_type = lfrx.boolean_element_type
    floating_point_type = lfrx.floating_point_element_types[0]

    condition = lfrx.as_field(
        lfrx.create_array((600, 400), lfrx.create_scalar(1, boolean_type))
    )
    x_coordinates = lfrx.cast(lfrx.cell_index(condition, 1), floating_point_type)
    y_coordinates = lfrx.cast(lfrx.cell_index(condition, 0), floating_point_type)
    result = lfrx.open_simplex_noise(x_coordinates, y_coordinates, seed=5)

    lfrx.to_gdal(result, str(result_array_path))
    # example]


if __name__ == "__main__":
    sys.exit(main())
